tool/tsh/proxy.go

```diff
@@ -145,94 +168,141 @@ func setupJumpHost(cf *CLIConf, tc *libclient.TeleportClient, sp sshProxyParams)
-func cleanTargetHost(targetHost, proxyHost, siteName string) string {
-	targetHost = strings.TrimSuffix(targetHost, "."+proxyHost)
-	targetHost = strings.TrimSuffix(targetHost, "."+siteName)
-	return targetHost
-}
+// sshProxy opens up a new SSH session connected to the Teleport Proxy's SSH proxy subsystem,
+// This is the equivalent of `ssh -o 'ForwardAgent yes' -p port %r@host -s proxy:%h:%p`.
+// If tls routing is enabled, the connection to RemoteProxyAddr is wrapped with TLS protocol.
+func sshProxy(ctx context.Context, tc *libclient.TeleportClient, sp sshProxyParams) error {
+	upstreamConn, err := dialSSHProxy(ctx, tc, sp)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer upstreamConn.Close()
 
-func sshProxyWithTLSRouting(cf *CLIConf, tc *libclient.TeleportClient, params sshProxyParams) error {
-	pool, err := tc.LocalAgent().ClientCertPool(params.clusterName)
+	client, err := makeSSHClient(ctx, upstreamConn, upstreamConn.RemoteAddr().String(), &ssh.ClientConfig{
+		User: tc.HostLogin,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeysCallback(tc.LocalAgent().Signers),
+		},
+		HostKeyCallback: tc.HostKeyCallback,
+	})
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	tlsConfig := &tls.Config{
-		RootCAs: pool,
+	defer client.Close()
+
+	sess, err := client.NewSession(ctx)
+	if err != nil {
+		return trace.Wrap(err)
 	}
+	defer sess.Close()
 
-	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
-		RemoteProxyAddr:    net.JoinHostPort(params.proxyHost, params.proxyPort),
-		Protocols:          []alpncommon.Protocol{alpncommon.ProtocolProxySSH},
-		InsecureSkipVerify: cf.InsecureSkipVerify,
-		ParentContext:      cf.Context,
-		SNI:                params.proxyHost,
-		SSHUser:            tc.HostLogin,
-		SSHUserHost:        fmt.Sprintf("%s:%s", params.targetHost, params.targetPort),
-		SSHHostKeyCallback: tc.HostKeyCallback,
-		SSHTrustedCluster:  params.clusterName,
-		ClientTLSConfig:    tlsConfig,
-	})
+	err = agent.ForwardToAgent(client.Client, tc.LocalAgent())
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	defer lp.Close()
-	if err := lp.SSHProxy(cf.Context, tc.LocalAgent()); err != nil {
+	err = agent.RequestAgentForwarding(sess)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	sshUserHost := fmt.Sprintf("%s:%s", sp.targetHost, sp.targetPort)
+	if err = sess.RequestSubsystem(proxySubsystemName(sshUserHost, sp.clusterName)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := proxySession(ctx, sess); err != nil {
 		return trace.Wrap(err)
 	}
 	return nil
 }
 
-func sshProxy(tc *libclient.TeleportClient, params sshProxyParams) error {
-	sshPath, err := getSSHPath()
+func dialSSHProxy(ctx context.Context, tc *libclient.TeleportClient, sp sshProxyParams) (net.Conn, error) {
+	remoteProxyAddr := net.JoinHostPort(sp.proxyHost, sp.proxyPort)
+
+	if !sp.tlsRouting {
+		conn, err := net.Dial("tcp", remoteProxyAddr)
```

Comment:
Now I noticed that. Here you could use "context aware" version of dial
```suggestion
		conn, err := (&net.Dialer{}).DialContext(ctx, "tcp", remoteProxyAddr)
```